### PR TITLE
feat!: move test runner into a separate package

### LINKED ISSUE #1470: Custom test runner
### Clear and concise description of the problem

Jest ecosystem has [jest-runner](https://github.com/facebook/jest/tree/main/packages/jest-runner) or [create-jest-runner](https://github.com/jest-community/create-jest-runner).

It can be extended to test in an environment such as an [electron](https://github.com/d4rkr00t/jest-electron-runner).
Also, It would be nice to be able to perform [lint](https://github.com/jest-community/jest-runner-eslint) or [typecheck](https://github.com/jest-community/jest-runner-tsd) with the efficient worker of tinypool.

### Suggested solution

I'd like to provide an API for setting up a custom runner.

### Alternative

It can also refer to Jest's [`projects` config](https://jestjs.io/docs/configuration#projects-arraystring--projectconfig).

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by sheremet-va**
This should make extending runners easier. This change also allows disabling auto inlining of the Vitest package (we only inline a single file that does "import").

Also browser package can now directly import runner.

Consequences of this PR:
- Userland can now import "vitest" directly in packages
- Snapshot testing is supported in the browser
- Tests are running ~20% faster, because we don't inline vitest anymore (I tested only test/core for this claim)
- Browser testing works on initial load, bot doesn't reload on restart
- Docs has "advanced" section with `startVitest` and new runner API
- @vitest/* packages are no longer bundled with Vitest
- Userland can define their own "test" function (like benchmark) that will be collected with `describe` and default `@vitest/runner`

Closes #1470

### CODE REVIEW

**Comment by sheremet-va** [comment]
@AriPerkkio coverage tests are failing again, I am not sure what can cause this here, can you have a look in your free time, please? 👀 

@Aslemammad @userquin I broke browser package, and not sure how you envision it with this change. The idea right now is to use `@vitest/runner` directly. Do you think you can fix it in this PR?

**Comment by AriPerkkio** [comment]
> coverage tests are failing again

Quick look shows that sourcemaps are a bit off.  Maybe the "magic number" given as offset to sourcemaps in c8 provider is now different? I can debug further in two days. 

It's good that there are now tests validating this, and failing for a real reason. 👍

**Comment by sheremet-va** [comment]
> > coverage tests are failing again
> 
> 
> 
> Quick look shows that sourcemaps are a bit off.  Maybe the "magic number" given as offset to sourcemaps in c8 provider is now different? I can debug further in two days. 
> 
> 
> 
> It's good that there are now tests validating this, and failing for a real reason. 👍 

Yes! You a probably right, I removed mocker from context. Thank you 🙏

**Comment by sheremet-va** [comment]
`--browser` is fixed. Also added possibility to use snapshot tests in browser 😄

**Comment by sheremet-va** [comment]
@antfu this is ready for review. Don't be scared by the numbers, most of it is lockfile 😄

**Comment by Aslemammad** [comment]
@sheremet-va That'd improve the experience of the browser, thank you so much Vladimir.

**Review by antfu — APPROVED** [review]
The pref improvement is sweet!

I am not able to review line by line 😅. I think this is a good move, and I trust your implementation 👍

File: docs/advanced/runner.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+```
+
+When initiating this class, Vitest passes down Vitest config, - you should expose it as a `config` property.
+
+::: warning
+`importFile` method in your custom runner must be inlined in `deps.inline` config option, if you call Node `import` inside.
+:::
+
+::: tip
+Snapshot support and some other features depend on the runner. If you don't want to lose it, you can extend your runner from `VitestTestRunner` imported from `vitest/runners`. It also exposes `BenchmarkNodeRunner`, if you want to extend benchmark functionality.
```
Comment (snyamathi) [inline docs/advanced/runner.md:93]: Is there a more concrete example of this?

```js
import { VitestRunner } from '@vitest/runner';
         ^^^^^^^^^^^^
SyntaxError: The requested module '@vitest/runner' does not provide an export named 'VitestRunner'
```

I don't see VitestRunner exported in https://unpkg.com/@vitest/runner@0.28.4/dist/index.js

**Comment by n3ur0n** [comment]
Hello @sheremet-va! What's the state of that PR? Is it possible to run a custom test runner, mentioned in #1470?

**Pull request merged**